### PR TITLE
feature: add step

### PR DESCRIPTION
**Pull request by ireznice**

### CODE REVIEW

File: lib/travis/guest-api/app/endpoint/steps.rb [hunk 1]
```diff
@@ -8,44 +8,58 @@ class Steps < Travis::GuestApi::App::Endpoint
     end
```
Comment (ondrej-hosak) [inline lib/travis/guest-api/app/endpoint/steps.rb]: pridat komentare

File: spec/integration/endpoints/steps_spec.rb [hunk 2]
```diff
@@ -123,6 +127,57 @@ def create_step(step)
...
+            {
+              'name'      => 'stepName2',
+              'classname' => 'caseName2',
+              'result'    => 'success',
+              'test_data' => { 'any_content' => 'xxx' },
+              'duration' => 56
+            }
+          }
+          let(:cache) {
+            Travis::GuestAPI::Cache.new 100 , Travis.config.redis
```
Comment (ondrej-hosak) [inline spec/integration/endpoints/steps_spec.rb]: Robocop

File: lib/travis/guest-api/app/endpoint/steps.rb [hunk 3]
```diff
@@ -8,44 +8,58 @@ class Steps < Travis::GuestApi::App::Endpoint
...
     post '/steps' do
-      steps = env['rack.parser.result']
-      steps = [ params ] unless (Array === env['rack.parser.result'])
+      begin
+        steps = env['rack.parser.result']
+        steps = [ params ] unless (Array === env['rack.parser.result'])
+
+        steps.map! do |step|
+
+          halt 422, {
```
Comment (ondrej-hosak) [inline lib/travis/guest-api/app/endpoint/steps.rb]: better to raise exception, because exception are handled in rescue code

File: lib/travis/guest-api/cache.rb [hunk 4]
```diff
@@ -60,6 +63,22 @@ def get_result(job_id)
       return passed ? 'passed' : 'failed'
     end
 
+    def get_added_step_metrics(job_id, class_name)
+      job_record = get_job(job_id)
+      raise AddStepException, "Invalid class_name: #{class_name}" if job_record.nil?
```
Comment (ondrej-hosak) [inline lib/travis/guest-api/cache.rb]: invalid class name exception when searching by id?

File: lib/travis/guest-api/cache.rb [hunk 5]
```diff
@@ -60,6 +63,22 @@ def get_result(job_id)
...
+      items = job_record.values.select { |obj| obj['classname'] == class_name }
```
Comment (ondrej-hosak) [inline lib/travis/guest-api/cache.rb]: class_index = job_record.values.select do |obj|
  obj['class_possition'].to_i if obj['classname'] == class_name
end.max

step_index = job_record.values.select do |obj|
  obj['possition'].to_i if obj['class_possition'] == class_index
end.max

File: lib/travis/guest-api/app/endpoint/steps.rb [hunk 6]
```diff
@@ -8,44 +8,59 @@ class Steps < Travis::GuestApi::App::Endpoint
...
+            'data',
+            'test_data',
+            'added_step'
+          )
+          res['number'] = 0
+          res
+        end
 
-      steps.map! do |step|
+puts "steps: #{steps}"
```
Comment (ondrej-hosak) [inline lib/travis/guest-api/app/endpoint/steps.rb]: rm

File: lib/travis/guest-api/app/endpoint/steps.rb (hunk 3 above)
Comment (panjan) [inline lib/travis/guest-api/app/endpoint/steps.rb]: I would agree with `halt 422` rather than work by exception. In addition, we could tighten the `begin-rescue` so that it only encloses the code which throws `AddStepException`.

File: lib/travis/guest-api/app/endpoint/steps.rb [hunk 7]
```diff
@@ -8,44 +8,59 @@ class Steps < Travis::GuestApi::App::Endpoint
...
-      steps = env['rack.parser.result']
-      steps = [ params ] unless (Array === env['rack.parser.result'])
+      begin
+        steps = env['rack.parser.result']
+        steps = [ params ] unless (Array === env['rack.parser.result'])
+
+        steps.map! do |step|
+
+          raise Travis::GuestAPI::Cache::AddStepException,
+            'Keys name, classname are mandatory!' unless !step.nil? &&
```
Comment (petrSchreiber) [inline lib/travis/guest-api/app/endpoint/steps.rb]: Just aesthetical suggestion:
"Mandatory parameters 'name' and 'classname' need to be specified'"

File: lib/travis/guest-api/app/endpoint/steps.rb [hunk 8]
```diff
@@ -8,44 +8,59 @@ class Steps < Travis::GuestApi::App::Endpoint
...
+              step['name'] && step['classname']
+
+          raise Travis::GuestAPI::Cache::AddStepException,
+            'UUID cannot be set!' if step['uuid']
```
Comment (petrSchreiber) [inline lib/travis/guest-api/app/endpoint/steps.rb]: Just aesthetical suggestion:
"The specified 'uuid' is already used"

File: lib/travis/guest-api/cache.rb [hunk 9]
```diff
@@ -60,6 +63,27 @@ def get_result(job_id)
       return passed ? 'passed' : 'failed'
     end
 
+    def get_added_step_metrics(job_id, class_name)
+      job_record = get_job(job_id)
+      raise AddStepException,
+        "Invalid cache content class_name: #{class_name} couldn\'t be searched" if job_record.nil?
```
Comment (petrSchreiber) [inline lib/travis/guest-api/cache.rb]: What do you think about this?:
"Test case #{class_name} could not be found for job id: #{job_id}"

File: spec/integration/endpoints/steps_spec.rb [hunk 10]
```diff
@@ -123,6 +127,57 @@ def create_step(step)
...
 
         end
+
+        context 'add_step functionality' do
+          let(:testcase_addstep) {
+            {
+              'name'      => 'stepName2',
+              'classname' => 'caseName2',
+              'result'    => 'success',
+              'test_data' => { 'any_content' => 'xxx' },
```
Comment (petrSchreiber) [inline spec/integration/endpoints/steps_spec.rb]: this seems like explicit_content :)

File: spec/integration/endpoints/steps_spec.rb [hunk 11]
```diff
@@ -123,6 +127,57 @@ def create_step(step)
...
+              'duration' => 56
+            }
+          }
+          let(:cache) {
+            Travis::GuestAPI::Cache.new 100 , Travis.config.redis
+          }
+          let(:sample_record) {
+            {
+              'name' => 'foo',
+              'classname' => 'caseName2',
```
Comment (petrSchreiber) [inline spec/integration/endpoints/steps_spec.rb]: I think this shows a bit of issue in naming convention. When we have class_position, then why not class_name as well?

File: lib/travis/guest-api/app/endpoint/steps.rb (hunk 3 above)
Comment (ondrej-hosak) [inline lib/travis/guest-api/app/endpoint/steps.rb]: +1

File: spec/integration/endpoints/steps_spec.rb [hunk 12]
```diff
@@ -123,6 +127,57 @@ def create_step(step)
...
+
+          it 'responds with 422 when redis cache is empty' do
+            response = post '/api/v2/steps', testcase_addstep.to_json, "CONTENT_TYPE" => "application/json"
+            expect(response.status).to eq(422)
+            expect(response.body).to match(/Invalid\ cache\ content/)
+          end
+
+          it 'responds with 200 when adding into existing class' do
+            cache.set 1, 1, sample_record
+            expect(reporter).to receive(:send_tresult) { |job_id, arg|
```
Comment (ondrej-hosak) [inline spec/integration/endpoints/steps_spec.rb]: unused block?

File: spec/integration/endpoints/steps_spec.rb [hunk 13]
```diff
@@ -196,10 +251,20 @@ def create_step(step)
 
         context 'bulk update' do
           let(:testcase1) {
-            { 'name' => 'stepName1', 'classname' =>  'testCaseName1' }
+            {
```
Comment (ondrej-hosak) [inline spec/integration/endpoints/steps_spec.rb]: bulk update works the same way, you can use shared examples.

File: spec/integration/endpoints/steps_spec.rb [hunk 14]
```diff
@@ -123,6 +127,57 @@ def create_step(step)
...
+
+          before :each do
+            @redis = Redis.new Travis.config.redis
+            @redis.flushdb
+          end
+
+          it 'responds with 422 when redis cache is empty' do
+            response = post '/api/v2/steps', testcase_addstep.to_json, "CONTENT_TYPE" => "application/json"
+            expect(response.status).to eq(422)
+            expect(response.body).to match(/Invalid\ cache\ content/)
```
Comment (ireznice) [inline spec/integration/endpoints/steps_spec.rb]: remove

File: lib/travis/guest-api/app/endpoint/steps.rb (hunk 8 above)
Comment (petrSchreiber) [inline lib/travis/guest-api/app/endpoint/steps.rb]: correction:  "'uuid' field is not allowed to be set"

File: spec/integration/endpoints/steps_spec.rb (hunk 11 above)
Comment (petrSchreiber) [inline spec/integration/endpoints/steps_spec.rb]: explained by ivor, this is not related to this PR
